Quick update: With the following crude patch to auto-evaluate `exp` I'm getting even more test failures... I didn't inspect them all, a few seemed rather benign. (slow tests not tested)
```
--- a/sympy/functions/elementary/exponential.py
+++ b/sympy/functions/elementary/exponential.py
@@ -263,6 +263,13 @@ def eval(cls, arg):
                             return -S.ImaginaryUnit
                         elif ask(Q.odd(coeff + S.Half)):
                             return S.ImaginaryUnit
+                    elif coeff.is_Rational:
+                        from sympy.functions.elementary.trigonometric \
+                                import cos, sin
+                        if coeff.q in [3, 4, 6]:
+                            I = S.ImaginaryUnit
+                            return cos(coeff*S.Pi) + I*sin(coeff*S.Pi)
```

```
______________________________________________ sympy/core/tests/test_arit.py:test_mul_coeff _______________________________________________
Traceback (most recent call last):
  File "/root/sympy/sympy/core/tests/test_arit.py", line 1896, in test_mul_coeff
    assert p**2*x*p*y*p*x*p**2 == x**2*y
AssertionError
___________________________________________________________________________________________________________________________________________
____________________________ sympy/physics/quantum/tests/test_identitysearch.py:test_bfs_identity_search_xfail ____________________________
Traceback (most recent call last):
  File "/root/sympy/sympy/physics/quantum/tests/test_identitysearch.py", line 497, in test_bfs_identity_search_xfail
    assert bfs_identity_search(gate_list, 1, max_depth=3) == id_set
AssertionError
___________________________________________________________________________________________________________________________________________
___________________________________ sympy/physics/quantum/tests/test_spin.py:test_represent_spin_states ___________________________________
Traceback (most recent call last):
  File "/root/sympy/sympy/physics/quantum/tests/test_spin.py", line 70, in test_represent_spin_states
    JzKet(S(1)/2, S(1)/2), basis=Jy) == sqrt(2)*Matrix([-1, I])/2
AssertionError
___________________________________________________________________________________________________________________________________________
________________________________ sympy/physics/quantum/tests/test_spin.py:test_represent_uncoupled_states _________________________________
Traceback (most recent call last):
  File "/root/sympy/sympy/physics/quantum/tests/test_spin.py", line 110, in test_represent_uncoupled_states
    Matrix([-I, 0, 0, 0])
AssertionError
___________________________________________________________________________________________________________________________________________
________________________________________ sympy/physics/quantum/tests/test_spin.py:test_rotation_d _________________________________________
Traceback (most recent call last):
  File "/root/sympy/sympy/physics/quantum/tests/test_spin.py", line 3595, in test_rotation_d
    S(1)/2, S(1)/2, S(1)/2, pi/2, pi/2, pi/2).doit() == -I*sqrt(2)/2
AssertionError
___________________________________________________________________________________________________________________________________________
________________________________________ sympy/simplify/tests/test_hyperexpand.py:test_branch_bug _________________________________________
Traceback (most recent call last):
  File "/root/sympy/sympy/simplify/tests/test_hyperexpand.py", line 25, in test_branch_bug
    + sqrt(pi)*erf(sqrt(z))/(5*sqrt(z))
AssertionError
___________________________________________________________________________________________________________________________________________
____________________________________________ sympy/simplify/tests/test_powsimp.py:test_powsimp ____________________________________________
Traceback (most recent call last):
  File "/root/sympy/sympy/simplify/tests/test_powsimp.py", line 99, in test_powsimp
    assert str( powsimp(exp(I*pi/3)*root(-1,3)) ) == '(-1)**(2/3)'
AssertionError
___________________________________________________________________________________________________________________________________________
________________________________________ sympy/simplify/tests/test_simplify.py:test_simplify_other ________________________________________
Traceback (most recent call last):
  File "/root/sympy/sympy/simplify/tests/test_simplify.py", line 155, in test_simplify_other
    assert simplify(ans) == -(-1)**(S(3)/4)*sqrt(pi)/sqrt(t)
AssertionError
___________________________________________________________________________________________________________________________________________
__________________________________________ sympy/simplify/tests/test_simplify.py:test_besselsimp __________________________________________
Traceback (most recent call last):
  File "/root/sympy/sympy/simplify/tests/test_simplify.py", line 603, in test_besselsimp
    besselj(a, sqrt(x)) * cos(sqrt(x))
AssertionError
```

--- a/sympy/functions/elementary/exponential.py
+++ b/sympy/functions/elementary/exponential.py
@@ -408,7 +408,7 @@ class exp(ExpBase):
     def _eval_nseries(self, x, n, logx):
         # NOTE Please see the comment at the beginning of this file, labelled
         #      IMPORTANT.
-        from sympy import limit, oo, Order, powsimp
+        from sympy import limit, oo, Order, powsimp, Wild, expand_complex
         arg = self.args[0]
         arg_series = arg._eval_nseries(x, n=n, logx=logx)
         if arg_series.is_Order:
@@ -423,7 +423,12 @@ class exp(ExpBase):
         r = exp(arg0)*exp_series.subs(t, arg_series - arg0)
         r += Order(o.expr.subs(t, (arg_series - arg0)), x)
         r = r.expand()
-        return powsimp(r, deep=True, combine='exp')
+        r = powsimp(r, deep=True, combine='exp')
+        # powsimp may introduce unexpanded (-1)**Rational; see PR #17201
+        simplerat = lambda x: x.is_Rational and x.q in [3, 4, 6]
+        w = Wild('w', properties=[simplerat])
+        r = r.replace((-1)**w, expand_complex((-1)**w))
+        return r
 
     def _taylor(self, x, n):
         from sympy import Order
@@ -582,37 +587,6 @@ class log(Function):
             elif arg is S.Exp1:
                 return S.One
 
-        if arg.is_number and arg.is_algebraic:
-            # Check for arguments involving rational
-            # multiples of pi
-            r_ = re(arg)
-            i_ = im(arg)
-            t = (i_/r_).cancel()
-            atan_table = {
-                # first quadrant only
-                sqrt(3): S.Pi/3,
-                1: S.Pi/4,
-                sqrt(2)*sqrt(5 - sqrt(5))/(1 + sqrt(5)): S.Pi/5,
-                sqrt(2)*sqrt(sqrt(5) + 5)/(-1 + sqrt(5)): 2*S.Pi/5,
-                sqrt(3)/3: S.Pi/6,
-                sqrt(2 - sqrt(2))/sqrt(sqrt(2) + 2): S.Pi/8,
-                sqrt(sqrt(2) + 2)/sqrt(2 - sqrt(2)): 3*S.Pi/8,
-                (-sqrt(2) + sqrt(10))/(2*sqrt(sqrt(5) + 5)): S.Pi/10,
-                (sqrt(2) + sqrt(10))/(2*sqrt(5 - sqrt(5))): 3*S.Pi/10,
-                (-1 + sqrt(3))/(1 + sqrt(3)): S.Pi/12,
-                (1 + sqrt(3))/(-1 + sqrt(3)): 5*S.Pi/12
-            }
-            if t in atan_table:
-                if r_.is_positive:
-                    return cls(Abs(arg)) + I * atan_table[t]
-                else:
-                    return cls(Abs(arg)) + I * (atan_table[t] - S.Pi)
-            elif -t in atan_table:
-                if r_.is_positive:
-                    return cls(Abs(arg)) + I * (-atan_table[-t])
-                else:
-                    return cls(Abs(arg)) + I * (S.Pi - atan_table[-t])
-
         # don't autoexpand Pow or Mul (see the issue 3351):
         if not arg.is_Add:
             coeff = arg.as_coefficient(S.ImaginaryUnit)
@@ -627,6 +601,57 @@ class log(Function):
                         return S.Pi * S.ImaginaryUnit * S.Half + cls(coeff)
                     else:
                         return -S.Pi * S.ImaginaryUnit * S.Half + cls(-coeff)
+
+        if arg.is_number and arg.is_algebraic:
+            I = S.ImaginaryUnit
+            r_, i_ = arg.as_independent(I, as_Add=True)
+            i_ = i_.as_coefficient(I)
+            if i_ and i_.is_real and r_.is_real:
+                if r_.is_zero:
+                    if i_.is_positive:
+                        return S.Pi * I * S.Half + cls(i_)
+                    elif i_.is_negative:
+                        return -S.Pi * I * S.Half + cls(-i_)
+                    elif i_.is_zero:
+                        return zoo
+                else:
+                    from sympy.simplify import ratsimp
+                    # Check for arguments involving rational multiples of pi
+                    t = (i_/r_).cancel()
+                    atan_table = {
+                        # first quadrant only
+                        sqrt(3): S.Pi/3,
+                        1: S.Pi/4,
+                        sqrt(5 - 2*sqrt(5)): S.Pi/5,
+                        sqrt(2)*sqrt(5 - sqrt(5))/(1 + sqrt(5)): S.Pi/5,
+                        sqrt(5 + 2*sqrt(5)): 2*S.Pi/5,
+                        sqrt(2)*sqrt(sqrt(5) + 5)/(-1 + sqrt(5)): 2*S.Pi/5,
+                        sqrt(3)/3: S.Pi/6,
+                        sqrt(2) - 1: S.Pi/8,
+                        sqrt(2 - sqrt(2))/sqrt(sqrt(2) + 2): S.Pi/8,
+                        sqrt(2) + 1: 3*S.Pi/8,
+                        sqrt(sqrt(2) + 2)/sqrt(2 - sqrt(2)): 3*S.Pi/8,
+                        sqrt(1 - 2*sqrt(5)/5): S.Pi/10,
+                        (-sqrt(2) + sqrt(10))/(2*sqrt(sqrt(5) + 5)): S.Pi/10,
+                        sqrt(1 + 2*sqrt(5)/5): S.Pi/10,
+                        (sqrt(2) + sqrt(10))/(2*sqrt(5 - sqrt(5))): 3*S.Pi/10,
+                        2 - sqrt(3): S.Pi/12,
+                        (-1 + sqrt(3))/(1 + sqrt(3)): S.Pi/12,
+                        2 + sqrt(3): 5*S.Pi/12,
+                        (1 + sqrt(3))/(-1 + sqrt(3)): 5*S.Pi/12
+                    }
+                    if t in atan_table:
+                        modulus = ratsimp(Abs(arg))
+                        if r_.is_positive:
+                            return cls(modulus) + I * atan_table[t]
+                        else:
+                            return cls(modulus) + I * (atan_table[t] - S.Pi)
+                    elif -t in atan_table:
+                        modulus = ratsimp(Abs(arg))
+                        if r_.is_positive:
+                            return cls(modulus) + I * (-atan_table[-t])
+                        else:
+                            return cls(modulus) + I * (S.Pi - atan_table[-t])
 
     def as_base_exp(self):
         """

--- a/sympy/functions/elementary/tests/test_exponential.py
+++ b/sympy/functions/elementary/tests/test_exponential.py
@@ -2,7 +2,7 @@ from sympy import (
     symbols, log, ln, Float, nan, oo, zoo, I, pi, E, exp, Symbol,
     LambertW, sqrt, Rational, expand_log, S, sign, conjugate, refine,
     sin, cos, sinh, cosh, tanh, exp_polar, re, Function, simplify,
-    AccumBounds, MatrixSymbol, Pow)
+    AccumBounds, MatrixSymbol, Pow, gcd)
 from sympy.abc import x, y, z
 from sympy.core.expr import unchanged
 from sympy.core.function import ArgumentIndexError
@@ -191,9 +191,22 @@ def test_log_values():
     assert log(0, 2) == zoo
     assert log(0, 5) == zoo
 
-    assert log(sqrt(3)/2 + I*S.Half) == I*pi/6
+    assert exp(-log(3))**(-1) == 3
+
+    assert log(S.Half) == -log(2)
+    assert log(2*3).func is log
+    assert log(2*3**2).func is log
+
+
+def test_log_exact():
+    # check for pi/2, pi/3, pi/4, pi/6, pi/8, pi/12; pi/5, pi/10:
+    for n in range(-23, 24):
+        if gcd(n, 24) != 1:
+            assert log(exp(n*I*pi/24).rewrite(sqrt)) == n*I*pi/24
+        for n in range(-9, 10):
+            assert log(exp(n*I*pi/10).rewrite(sqrt)) == n*I*pi/10
+
     assert log(S.Half - I*sqrt(3)/2) == -I*pi/3
-    assert log(sqrt(2)/2 + I*sqrt(2)/2) == I*pi/4
     assert log(-S.Half + I*sqrt(3)/2) == I*2*pi/3
     assert log(-sqrt(2)/2 - I*sqrt(2)/2) == -I*3*pi/4
     assert log(-sqrt(3)/2 - I*S.Half) == -I*5*pi/6
@@ -205,14 +218,18 @@ def test_log_values():
 
     assert log(-1 + I*sqrt(3)) == log(2) + I*2*pi/3
     assert log(5 + 5*I) == log(5*sqrt(2)) + I*pi/4
-    assert log(5*(1 + I)/sqrt(2)) == log(5) + I*pi/4
+    assert log(sqrt(-12)) == log(2*sqrt(3)) + I*pi/2
     assert log(-sqrt(6) + sqrt(2) - I*sqrt(6) - I*sqrt(2)) == log(4) - I*7*pi/12
+    assert log(1 + I*sqrt(2-sqrt(2))/sqrt(2+sqrt(2))) == log(2/sqrt(sqrt(2) + 2)) + I*pi/8
+    assert log(cos(7*pi/12) + I*sin(7*pi/12)) == 7*I*pi/12
+    assert log(cos(6*pi/5) + I*sin(6*pi/5)) == -4*I*pi/5
 
-    assert exp(-log(3))**(-1) == 3
+    zero = (1 + sqrt(2))**2 - 3 - 2*sqrt(2)
+    assert log(zero - I*sqrt(3)) == log(sqrt(3)) - I*pi/2
+    assert unchanged(log, zero + I*zero) or log(zero + zero*I) == zoo
 
-    assert log(S.Half) == -log(2)
-    assert log(2*3).func is log
-    assert log(2*3**2).func is log
+    # bail quickly if no obvious simplification is possible:
+    assert unchanged(log, (sqrt(2)-1/sqrt(sqrt(3)+I))**1000)
 
 
 def test_log_base():

